micrometer-shim/src/main/java/io/opentelemetry/contrib/metrics/micrometer/DoubleCounter.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package io.opentelemetry.contrib.metrics.micrometer;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.Tags;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.metrics.BoundDoubleCounter;
+import io.opentelemetry.api.metrics.DoubleCounterBuilder;
+import io.opentelemetry.api.metrics.LongCounterBuilder;
+import io.opentelemetry.api.metrics.ObservableDoubleMeasurement;
+import io.opentelemetry.context.Context;
+import java.util.function.Consumer;
+
+final class DoubleCounter implements io.opentelemetry.api.metrics.DoubleCounter {
+  private final SharedMeterState state;
+  private final Reader<Counter.Builder> factory;
+
+  private DoubleCounter(SharedMeterState state, Reader<Counter.Builder> factory) {
+    this.state = state;
+    this.factory = factory;
+  }
+
+  @Override
+  public void add(double value) {
+    add(value, Attributes.empty(), Context.current());
+  }
+
+  @Override
+  public void add(double value, Attributes attributes) {
+    add(value, attributes, Context.current());
+  }
+
+  @Override
+  public void add(double value, Attributes attributes, Context context) {
+    BoundDoubleCounter bound = bind(attributes);
+    try {
+      bound.add(value, context);
+    } finally {
+      bound.unbind();
+    }
+  }
+
+  @Override
+  public BoundDoubleCounter bind(Attributes attributes) {
+    Tags tags = TagUtils.attributesToTags(attributes);
+    Counter counter = factory.get().tags(tags).register(state.meterRegistry());
+    return new Bound(counter);
+  }
+
+  static DoubleCounterBuilder newBuilder(SharedMeterState state, String name) {
+    return new Builder(state, () -> Counter.builder(name));
+  }
+
+  static DoubleCounterBuilder newBuilder(SharedMeterState state, Reader<Counter.Builder> factory) {
+    return new Builder(state, factory);
+  }
+
+  static final class Builder implements DoubleCounterBuilder {
+    private final SharedMeterState state;
+    private final Reader<Counter.Builder> factory;
+
+    Builder(SharedMeterState state, Reader<Counter.Builder> factory) {
+      this.state = state;
+      this.factory = factory;
+    }
+
+    @Override
+    public Builder setDescription(String description) {
+      return new Builder(state, factory.andThen(builder -> builder.description(description)));
+    }
+
+    @Override
+    public Builder setUnit(String unit) {
+      return new Builder(state, factory.andThen(builder -> builder.baseUnit(unit)));
+    }
+
+    @Override
+    public LongCounterBuilder ofLongs() {
+      return LongCounter.newBuilder(state, factory);
+    }
+
+    @Override
+    public DoubleCounter build() {
+      return new DoubleCounter(state, factory);
+    }
+
+    @Override
+    public void buildWithCallback(Consumer<ObservableDoubleMeasurement> callback) {
```

Comment:
This implementation means the async measurements are recorded *ONCE* and then never again.

Reply:
Thanks for the heads up, that's my misunderstanding of the OTel API.  To confirm, `callback` is then called once per export?

Reply:
Yeah.  The idea is that anytime you export, this callback should be called and you synthesize your metrics from it right then.